src/jit/optimizer.cpp

```diff
@@ -6795,8 +6795,22 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
 
         bool IsTreeVNInvariant(GenTree* tree)
         {
-            return m_compiler->optVNIsLoopInvariant(tree->gtVNPair.GetLiberal(), m_loopNum,
-                                                    &m_hoistContext->m_curLoopVnInvariantCache);
+            ValueNum vn = tree->gtVNPair.GetLiberal();
+
+            if (m_compiler->vnStore->IsVNConstant(vn))
+            {
+                // We only allow true constant nodes  (i.e. GT_CNS_INT, etc...) or a GT_ARGPLACE nodes.
+                // It is unsafe to allow a GT_CLS_VAR that has been assigned a constant.
+                // The logic in optVNIsLoopInvariant woudl consider it to be loop-invariant, even
+                // if the assignment of the constant to the GT_CLS_VAR was inside the loop.
+                //
+                if (!tree->OperIsConst() && tree->OperGet() != GT_ARGPLACE)
```

Comment:
This would be cleaner as 
```
if (!tree->OperIsConst() && !tree->OperIs(GT_ARGPLACE))
```
But if using the inequality, it should be parenthesized.